prefect_dask/task_runners.py

```diff
@@ -294,8 +295,18 @@ async def _start(self, exit_stack: AsyncExitStack):
             self._connect_to = self._cluster = await exit_stack.enter_async_context(
                 self.cluster_class(asynchronous=True, **self.cluster_kwargs)
             )
+
+            # If used with the operator implementation of KubeCluster,
+            # the cluster is not automatically started
+            if self._cluster.status.value != "running":
+                await self._cluster._start()
+
             if self.adapt_kwargs:
-                self._cluster.adapt(**self.adapt_kwargs)
+                # Depending on the cluster type (Cluster or SpecCluster),
+                # adapt should or shouldn't be awaited
+                adapt_response = self._cluster.adapt(**self.adapt_kwargs)
+                if iscoroutine(adapt_response):
+                    await adapt_response
```

Comment:
This will block the event loop if it's not a coroutine. You'll need to use `inspect.iscoroutinefn` then run it in a  worker thread if it is not (or check if its a `Cluster` / `SpecCluster`). This seems like an implementation issue in Dask though?

Reply:
I tried to use `inspect.iscoroutinefunction` on the `adapt` method itself and `asyncio.isfuture` on the return value but this method is [not declared async](https://github.com/dask/dask-kubernetes/blob/main/dask_kubernetes/operator/kubecluster/kubecluster.py#L597) and instead returns a future of the private `_adapt` handled with the method [sync](https://github.com/dask/distributed/blob/main/distributed/utils.py#L327).

Also, I am not sure to understand why we should run it in a worker thread if it's not async. I was trying to keep the same call as the original code [here](https://github.com/PrefectHQ/prefect-dask/blob/main/prefect_dask/task_runners.py#L298) when used with the `classic.KubeCluster` or other `cloudprovider.*Instance`.

Reply:
If you're running on an event loop and have some code that performs IO it must either be async and awaited or sync and run in a worker thread. The event loop is designed to switch quickly between a bunch of different tasks. When you run sync code in the event loop, it cannot switch to other tasks. This "blocks" the loop and causes performance problems.

It looks like the `sync` utility checks if `asynchronous` is set and returns a coroutine (they call it a "future", but it should be a coroutine from looking at the code).